src/replication/replication_node_manager.cpp

```diff
@@ -112,5 +165,47 @@ namespace cubreplication
       assert (g_slave_node != NULL);
       return g_slave_node;
     }
+
+    void wait_commute (HA_SERVER_STATE &ha_state, HA_SERVER_STATE req_state)
+    {
+      std::unique_lock<std::mutex> ul (g_commute_mtx);
+      g_commute_cv.wait (ul, [req_state, &ha_state] ()
+      {
+	return ha_state == req_state;
+      });
+    }
+
+    static void inc_ha_tasks_without_lock ()
+    {
+      ++g_ha_tasks_running;
+    }
+
+    void inc_ha_tasks ()
+    {
+      std::lock_guard<std::mutex> lg (g_ha_tasks_running_mtx);
+      inc_ha_tasks_without_lock ();
+    }
+
+    void dec_ha_tasks ()
+    {
+      std::lock_guard<std::mutex> lg (g_ha_tasks_running_mtx);
+      assert (g_ha_tasks_running > 0);
+      --g_ha_tasks_running;
+      if (g_ha_tasks_running == 0)
+	{
+	  g_ha_tasks_running_cv.notify_all ();
+	}
+    }
+
+    static std::unique_lock<std::mutex> wait_ha_tasks ()
```

Comment:
Regarding the solution of method returning the lock:
after we talked about the alternatives, I can understand the solution, but if I would have seen it for the first time, I would have needed a significant effort to understand it, compared with a classical solution (two methods : f and f_without_lock which are called as needed)

Reply:
Actually my case is a little delicate, I need to keep the lock I wait on after the return. I have the options: 1. return it, 2. pass it by reference or 3. inline wait_ha_tasks code at call-sites. I preferred the first option.